src/policy/discount.h

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2009-2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_POLICY_DISCOUNT_H
+#define BITCOIN_POLICY_DISCOUNT_H
+
+#include <consensus/consensus.h>
+#include <cstdint>
+#include <primitives/transaction.h>
+#include <version.h>
+
+/**
+ * Calculate a smaller virtual size for discounted Confidential Transactions.
+ */
+static inline int64_t GetDiscountVirtualTransactionSize(const CTransaction& tx, int64_t nSigOpCost = 0, unsigned int bytes_per_sig_op = 0)
+{
+    int64_t size_bytes = ::GetSerializeSize(tx, PROTOCOL_VERSION | SERIALIZE_TRANSACTION_NO_WITNESS) * (WITNESS_SCALE_FACTOR - 1) + ::GetSerializeSize(tx, PROTOCOL_VERSION);
+    int64_t sigop_bytes = nSigOpCost * bytes_per_sig_op;
+
+    int64_t weight = std::max(size_bytes, sigop_bytes);
+
+    // for each confidential output
+    for (size_t i = 0; i < tx.vout.size(); ++i) {
+        const CTxOut& output = tx.vout[i];
+        if (output.IsFee()) {
+            continue;
+        } else if (output.IsConfidential()) {
+            // subtract the weight of the output witness
+            weight -= ::GetSerializeSize(tx.witness.vtxoutwit[i], PROTOCOL_VERSION);
```

Comment:
I'm afraid you are going to need to wrap this in an `if (i < tx.witness.vtxoutwit.size())`.

Reply:
Great catch, thanks. Done.

Reply:
In 8bafd0c9fa448c3790ed2992f02b8e1245e6c56a:

Still not done (it may be done in a later commit, I am not there yet)

Reply:
Done in e24bc12bf572204f0b82eb360c78ce05e181dc33